### PR TITLE
pairwise difference outlier detection

### PR DESCRIPTION
**Pull request by Kadrinizar**
## Summary by Sourcery

Add a pairwise-difference-based outlier detection estimator with threshold optimization, update dependency constraints, and include a test script comparing it to IsolationForest.

New Features:
- Implement PairwiseDifferenceOutlierDetection estimator employing pairwise feature differences and automatic F1-based threshold selection.

Build:
- Loosen numpy version requirement, pin tqdm, and remove fastparquet from dependencies.

Tests:
- Add unsupervised_outlier_detection script to benchmark PairwiseDifferenceOutlierDetection against IsolationForest on real datasets.

### CODE REVIEW

File: pdll/_pairwise.py [hunk 1]
```diff
@@ -1,23 +1,57 @@
 """Pairwise Difference Learning meta-estimator."""
-import functools
-import warnings
-from typing import Iterable
+# Data manipulation and analysis
```
Comment (Karim-53) [inline pdll/_pairwise.py]:
```suggestion
```

File: pdll/_pairwise.py [hunk 2]
```diff
@@ -1,23 +1,56 @@
...
+warnings.filterwarnings('ignore')
+
+# Typing and utilities
+import functools
+from typing import Iterable
+
+# Author: Mohamed Karim Belaid <karim.belaid@idiada.com> or <extern.karim.belaid@porsche.de>
+# PairwiseDifferenceOutlierDetection class: Author Nizar Kadri <nizar.kadri@campus.lmu.de>
+# License: Apache-2.0 clause
+
```
Comment (Karim-53) [inline pdll/_pairwise.py:52]: move to line 2

File: pdll/_pairwise.py [hunk 3]
```diff
@@ -1,23 +1,56 @@
...
-from typing import Iterable
+import pandas as pd
+import numpy as np
 
-# Author: Mohamed Karim Belaid <karim.belaid@idiada.com> or <extern.karim.belaid@porsche.de>
-# License: Apache-2.0 clause
+# Data visualization
+import matplotlib.pyplot as plt
+import seaborn as sns
+
```
Comment (Karim-53) [inline pdll/_pairwise.py:8]: don't do import that are not used -.-'
same in the other .py
you can check that with VS code or with Copilot

File: pdll/_pairwise.py [hunk 4]
```diff
@@ -942,3 +975,141 @@ def transform(self, X, y=None):
...
+    """
+    Initializes the PairwiseDifferenceOutlierDetection model.
+
+    Parameters:
+    - estimator: A scikit-learn compatible estimator (default is IsolationForest).
+    """
+    self.estimator = estimator if estimator is not None else IsolationForest()
+    self.classifier = self.estimator
+
+def fit(self, X, y):
```
Comment (Karim-53) [inline pdll/_pairwise.py:996]: Since this is an unsupervised outlier detection y=None should be the default option

File: tests/unsupervised_outlier_detection.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+# Machine Learning
+from sklearn.metrics import roc_auc_score, average_precision_score
+from sklearn.datasets import fetch_openml
+from sklearn.model_selection import train_test_split
+from sklearn.preprocessing import StandardScaler
+from sklearn.neighbors import KNeighborsClassifier
+from sklearn.linear_model import LogisticRegression
+from sklearn.neighbors import NearestNeighbors
+from sklearn.ensemble import IsolationForest
+from sklearn.datasets import fetch_openml
```
Comment (Karim-53) [inline tests/unsupervised_outlier_detection.py:19]: try to shorten the imports to what you really need

File: tests/unsupervised_outlier_detection.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+# Ignore warnings for cleaner output
+import warnings
+warnings.filterwarnings('ignore')
+
+
+
+#  Dictionary of dataset file paths
+file_paths = {
+    1: "/datasets/celeba_baldvsnonbald_normalised.csv",
+    2: "/datasets/census-income-full-mixed-binarized.csv",
```
Comment (Karim-53) [inline tests/unsupervised_outlier_detection.py:39]: this is still missing in the repo (you did not upload it)

File: requirements.txt [hunk 7]
```diff
@@ -1,10 +1,10 @@
...
 scikit-learn~=1.3.2
-tqdm
+tqdm~=4.66.2
 openml~=0.14.2
 psutil~=5.9.8
 joblib~=1.2.0
 scipy~=1.11.4
-fastparquet
\ No newline at end of file
+fastparquet~=2024.2.0
```
Comment (Karim-53) [inline requirements.txt:10]: no. revert changes

File: pdll/_pairwise.py [hunk 8]
```diff
@@ -942,3 +975,141 @@ def transform(self, X, y=None):
...
+    """
+    self.estimator.fit(X)
+    self.X_train = X
+
+    # Create pairwise differences for training
+    X_train_pair, _ = self.__pair_input(X, X)
+    self.estimator.fit(X_train_pair)
+
+    # Calculate anomaly scores on the training pairs
+    scores = abs(self.estimator.score_samples(X_train_pair))
```
Comment (Karim-53) [inline pdll/_pairwise.py:1012]: don’t take abs, keep raw scores, and treat lower as more abnormal. Also, since you subclass OutlierMixin, align with scikit-learn convention: predict should return 1 for inliers and -1 for outliers.

Comment (Karim-53) [reply pdll/_pairwise.py:1012]:
```
- scores = abs(self.estimator.score_samples(X_train_pair))
+ scores = self.estimator.score_samples(X_train_pair)
  score_df = pd.DataFrame(scores.reshape((-1, len(self.X_train))))
  mean_scores = score_df.mean(axis=1).to_numpy()

- y_pred = np.where(mean_scores < threshold, 0, 1)
+ # 1 = inlier, -1 = outlier to match sklearn
+ y_pred = np.where(mean_scores < threshold, -1, 1)
```

Comment (Karim-53) [reply pdll/_pairwise.py:1012]:
```
def predict(self, X):
-    scores = self.score_samples(X)
-    return np.where(scores < self.threshold_, 0, 1)
+    scores = self.score_samples(X)
+    return np.where(scores < self.threshold_, -1, 1)

```

File: pdll/_pairwise.py [hunk 9]
```diff
@@ -942,3 +975,141 @@ def transform(self, X, y=None):
...
+    self.classifier = self.estimator
+
+def fit(self, X, y):
+    """
+    Fits the Isolation Forest model on the training data and finds the best threshold.
+
+    - Pairs all training samples (X_train × X_train)
+    - Trains the Isolation Forest on pairwise differences
+    - Searches for the best percentile threshold (1% to 99%) that maximizes F1-macro score
+    """
```
Comment (Karim-53) [inline pdll/_pairwise.py:1003]: set common sklearn attrs:
```
+ from sklearn.utils.validation import check_array, check_is_fitted
+ X = check_array(X, force_all_finite="allow-nan")
+ self.n_features_in_ = X.shape[1]
+ self.feature_names_in_ = getattr(X, "columns", None)
```

File: pdll/_pairwise.py [hunk 10]
```diff
@@ -942,3 +975,141 @@ def transform(self, X, y=None):
...
+    self.estimator.fit(X)
+    self.X_train = X
+
+    # Create pairwise differences for training
+    X_train_pair, _ = self.__pair_input(X, X)
+    self.estimator.fit(X_train_pair)
```
Comment (Karim-53) [inline pdll/_pairwise.py:1009]: why two .fit() ?????

File: tests/unsupervised_outlier_detection.py [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+from sklearn.base import BaseEstimator, OutlierMixin
+from sklearn.utils.validation import check_array, check_is_fitted
+from scipy.stats import sem, ttest_rel
+# Evaluation Metrics
+from sklearn.metrics import confusion_matrix, classification_report, accuracy_score, f1_score
+from pdll._pairwise import PairwiseDifferenceOutlierDetection
+
+# Ignore warnings for cleaner output
+import warnings
+warnings.filterwarnings('ignore')
```
Comment (Karim-53) [inline tests/unsupervised_outlier_detection.py:32]:
```suggestion
```

File: pdll/_pairwise.py [hunk 12]
```diff
@@ -942,3 +975,141 @@ def transform(self, X, y=None):
...
+
+def __init__(self, estimator=None):
+    """
+    Initializes the PairwiseDifferenceOutlierDetection model.
+
+    Parameters:
+    - estimator: A scikit-learn compatible estimator (default is IsolationForest).
+    """
+    self.estimator = estimator if estimator is not None else IsolationForest()
+    self.classifier = self.estimator
```
Comment (Karim-53) [inline pdll/_pairwise.py:994]: this is never used delete it

```suggestion
```

File: pdll/_pairwise.py [hunk 13]
```diff
@@ -942,3 +975,141 @@ def transform(self, X, y=None):
...
+
+    Returns:
+    - np.ndarray: Mean anomaly scores per sample.
+    """
+    X_test_pair, _  = self.__pair_input(X, self.X_train)
+    scores = abs(self.estimator.score_samples(X_test_pair))
+    score_df = pd.DataFrame(scores.reshape((-1, len(self.X_train))))
+    return score_df.mean(axis=1).to_numpy()
+
+def decision_function(self, X):
```
Comment (Karim-53) [inline pdll/_pairwise.py:1068]: Make it consistent with scikit-learn: positive = more normal.

File: tests/unsupervised_outlier_detection.py [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+        "Percentile": model.percentile_,  # The best percentile chosen during training
+        "F1_macro": f1_pdl,
+        "AUC_ROC": auc_roc_pdl,
+        "AUC_PR": auc_pr_pdl
+    })
+
+# Save all results
+df_results = pd.DataFrame(all_results)
+print("Summary:")
+print(df_results)
```
Comment (Karim-53) [inline tests/unsupervised_outlier_detection.py:124]: add assert on the expected performance
for example
assert F1_macro > 0.7 # the exact value should be arround 0.71

File: tests/unsupervised_outlier_detection.py [hunk 15]
```diff
@@ -0,0 +1,124 @@
+# Data manipulation and analysis
```
Comment (Karim-53) [inline tests/unsupervised_outlier_detection.py:1]: rename this file and move it to root 
`run_benchmark_outlier_detection.py`

then add a simple example in folder `examples/` and if you want , some unit tests in `tests`

**Comment by Karim-53** [comment]
@codex propose a fix to the failing github actions

File: pdll/_pairwise.py (hunk 4 above)
Comment (Kadrinizar) [reply pdll/_pairwise.py:996]: Since this is an unsupervised outlier detection y=None should be the default option:


if y is not None:
            # supervised mode → F1-Optimierung
            candidate_percentiles = np.arange(1, 100, 1)
            best_f1, best_threshold, best_percentile = -np.inf, None, None

            for p in candidate_percentiles:
                threshold = np.percentile(mean_scores, p)
                y_pred = np.where(mean_scores < threshold, 0, 1)
                f1 = f1_score(y, y_pred, average='macro')
                if f1 > best_f1:
                    best_f1, best_threshold, best_percentile = f1, threshold, p

            self.threshold_ = best_threshold
            self.percentile_ = best_percentile
            print(f"[supervised] Best percentile={best_percentile}, F1={best_f1:.4f}")
        else:
            # unsupervised mode → Threshold aus contamination
            perc = 100 * contamination
            self.threshold_ = np.percentile(mean_scores, perc)
            self.percentile_ = perc
            print(f"[unsupervised] Threshold set at {perc:.1f}th percentile")

        return self
        
I have extended the method so that it also works in a fully unsupervised manner —
that is, without using any label information.
However, it turned out that the results (F1, AUC) became worse,
since F1-based threshold tuning is no longer possible.
I am currently working on testing more adaptive thresholding strategies
to improve the performance of the unsupervised approach.

Dataset 1: celeba_baldvsnonbald_normalised.csv
[unsupervised] Threshold set at 10.0th percentile
 Dataset 2: bank-additional-full_normalised.csv
[unsupervised] Threshold set at 10.0th percentile
Summary:
     Dataset                              Method Percentile  F1_macro  \
0  Dataset 1                     IsolationForest   baseline  0.498930   
1  Dataset 1  PairwiseDifferenceOutlierDetection       10.0  0.106228   
2  Dataset 2                     IsolationForest   baseline  0.620101   
3  Dataset 2  PairwiseDifferenceOutlierDetection       10.0  0.114761   

AUC_ROC    AUC_PR  
0  0.601691  0.074599  
1  0.350866  0.045190  
2  0.631384  0.133820  
3  0.353317  0.054382
